bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/scheduling/Scheduler.java

```diff
@@ -465,6 +473,45 @@ public void poison() {
             runnableList.add(POISON_PILL);
         }
     }
+
+    public void setListenerDeclarationFound(boolean listenerDeclarationFound) {
+        this.listenerDeclarationFound = listenerDeclarationFound;
+        if (listenerDeclarationFound) {
+            setImmortal(true);
+        }
+    }
+
+    public boolean isListenerDeclarationFound() {
+        return listenerDeclarationFound;
+    }
+
+    public ListenerRegistry getListenerRegistry() {
+        return listenerRegistry;
+    }
+
+    /**
+     * The registry for runtime dynamic listeners.
+     */
+    public class ListenerRegistry {
+        private final Set<BObject> listenerSet = new HashSet<>();
+
+        public synchronized void registerListener(BObject listener) {
+            listenerSet.add(listener);
+            setImmortal(true);
+        }
+
+        public synchronized void deregisterListener(BObject listener) {
+            if (!isListenerDeclarationFound() && listenerSet.remove(listener) && listenerSet.isEmpty()) {
+                setImmortal(false);
```

Comment:
`listenerSet.remove(listener)` only invoked if `isListenerDeclarationFound` is false, is that the correct thing?

Reply:
Yes because if there's a module listener declared by the `listener` keyword is present we shouldn't set `immortal` to false even if `listenerSet` is empty.

Reply:
will fix